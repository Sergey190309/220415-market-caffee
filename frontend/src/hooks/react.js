import {useState, useEffect, useRef, useCallback } from 'react'

export const useAppState = useState
export const useAppEffect = useEffect
export const useAppRef = useRef
export const useAppCallback = useCallback
