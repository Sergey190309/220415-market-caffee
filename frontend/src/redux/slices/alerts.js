import { createSlice } from '@reduxjs/toolkit'

export const initialState = {
  /**
   * state structure:
   * alerts: [
   *   {
   *     message: 'mockMessage',
   *     alertType: 'info',
   *     timeout: 3000,
   *     id: 'mockId'
   *   }
   * ]
   */
  alerts: []
}

const alertsSlice = createSlice({
  name: 'alerts',
  initialState,
  reducers: {
    startAlert: (state, { payload }) => {
      state.alerts.push(payload)
    },
    removeAlert: (state, { payload }) => {
      // console.log('alertsSlicer, removeAlert, payload ->', payload);
      state.alerts.forEach((alert, index) => {
        if (alert.id === payload) {
          state.alerts.splice(index, 1)
        }
      })
    }
  }
})

export const { startAlert, removeAlert } = alertsSlice.actions
export const alertsSelector = state => state.alerts
export default alertsSlice.reducer
