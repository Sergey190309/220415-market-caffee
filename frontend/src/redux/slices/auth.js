import { createSlice } from '@reduxjs/toolkit'

import { axiosCommonToken, authAxiosClient } from '../../api/apiClient'
import { LOG_IN_INFO } from '../constants/localStorageVariables'

export const notLoggedInfo = {
  /**
   * Used for create and return to 'empty' set of state values
   */
  user_name: '',
  email: '',
  isAdmin: false,
  access_token: '',
  refresh_token: ''
}

export const logInInfo = () => {
  /**
   * Used to avoin undefined value if local ctorage does not contain
   * needed variables.
   */
  const _localStorage = localStorage.getItem(LOG_IN_INFO)
    ? { ...JSON.parse(localStorage.getItem(LOG_IN_INFO)), isLoggedIn: true }
    : { ...notLoggedInfo, isLoggedIn: false }
  // console.log('auth slice, logInInfo, _localStorage ->', _localStorage);
  return _localStorage
}

export const initialState = () => ({
  ...logInInfo(),
  loading: false,
  isSignedUp: false
})

const authSlice = createSlice({
  /**
   * I tested all reducers but signUpModalClosed and logInModalClosed.
   */
  name: 'auth',
  initialState: initialState(),
  reducers: {
    setState: (state, { payload }) => {
      // Object.keys(state).forEach(key => {
      //   delete state[key]
      // })
      Object.assign(state, payload)
    },
    signUpStart: state => {
      state.loading = true
      state.isSignedUp = false
      state.isLoggedIn = false
    },
    signUpSuccess: state => {
      localStorage.removeItem(LOG_IN_INFO)
      Object.assign(state, notLoggedInfo, {
        loading: false,
        isSignedUp: true,
        isLoggedIn: false
      })
    },
    signUpFail: state => {
      localStorage.removeItem(LOG_IN_INFO)
      Object.assign(state, notLoggedInfo, {
        isSignedUp: false,
        isLoggedIn: false,
        loading: false
      })
    },
    logInStart: state => {
      state.loading = true
      state.isSignedUp = false
      state.isLoggedIn = false
    },
    logInSuccess: (state, { payload }) => {
      // payload.isAuthenticated = true;
      // console.log('authSlice, logInSuccess, payload ->', payload);
      axiosCommonToken(payload.access_token, authAxiosClient)
      localStorage.setItem(LOG_IN_INFO, JSON.stringify(payload))
      Object.assign(state, payload, { loading: false, isLoggedIn: true })
    },
    logInFail: state => {
      localStorage.removeItem(LOG_IN_INFO)
      Object.assign(state, notLoggedInfo, {
        loading: false,
        isSignedUp: false,
        isLoggedIn: false
      })
    },
    logOut: state => {
      // console.log('authSlice, logOut')
      localStorage.removeItem(LOG_IN_INFO)
      Object.assign(state, notLoggedInfo, {
        loading: false,
        isSignedUp: false,
        isLoggedIn: false
      })
    },
    signUpModalClosed: state => {
      state.isSignedUp = false
    },
    logInModalClosed: state => {
      state.isLoggedIn = false
    }
  }
})

export const {
  setState,
  signUpStart,
  signUpSuccess,
  signUpFail,
  logInStart,
  logInSuccess,
  logInFail,
  logOut,
  signUpModalClosed,
  logInModalClosed
} = authSlice.actions

export const authSelector = state => state.auth

export default authSlice.reducer
