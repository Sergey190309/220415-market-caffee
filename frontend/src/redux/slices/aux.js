import { createSlice } from '@reduxjs/toolkit'

/**
 * Slice contains info for intern auxiliary actions
 * such as:
 * 1. Saving something to back end after password confirmation.
<<<<<<< HEAD
=======
 * I've postponed this course I decided to create other slice
 * to update back end witn new infor for admin.
>>>>>>> dev210823
 */

export const initialState = {
  savingItemsIdentity: {
    identity: '',
    view_id: '',
    content: ''
    // locale_id: ''
  }
}

const auxSlice = createSlice({
  name: 'aux',
  initialState,
  reducers: {
    setTestState: (state, { payload }) => {
      /**
       * That's for testing only.
       * It rewrites all state.
       */
      Object.assign(state, payload)
    }

  }
})

export const {
  setTestState
} = auxSlice.actions
export const auxSelector = state => state.aux
export default auxSlice.reducer
