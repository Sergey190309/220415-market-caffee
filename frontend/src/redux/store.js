import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit'
import createSagaMiddleware from 'redux-saga'
// import reducer from './reducers';
import reducer from './slices'
import rootSaga from './saga'
// import { rootReducer } from './reducers';

const initialState = {}
const sagaMiddleware = createSagaMiddleware()

const middleware = [...getDefaultMiddleware({ thunk: false }), sagaMiddleware]
// const middleware = [sagaMiddleware, thunk];

const store = configureStore({
  reducer,
  middleware,
  initialState
})

sagaMiddleware.run(rootSaga)

export default store
