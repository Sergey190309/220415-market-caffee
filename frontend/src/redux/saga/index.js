import { all } from 'redux-saga/effects'
import { logInSaga, signUpSaga } from './auth'
import { alertSaga } from './alerts'
import { startInitSaga, techInSaga, lngsSaga, i18nSaga } from './tech'
import { structureSaga } from './structure'

export default function * rootSaga () {
  yield all([
    alertSaga(),
    logInSaga(),
    signUpSaga(),
    startInitSaga(),
    techInSaga(),
    lngsSaga(),
    i18nSaga(),
    structureSaga()
  ])
}
