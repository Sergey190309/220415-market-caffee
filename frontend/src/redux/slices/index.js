import { combineReducers } from '@reduxjs/toolkit'
// import { combineReducers } from 'redux';
// import alerts from '../reducers/alert';
import alerts from './alerts'
// import logIn from '../reducers/auth';
import auth from './auth'
import device from './device'
// import device from '../reducers/device';
import lng from './lng'
// import lng from '../reducers/lng';
import structure from './structure'
import tech from './tech'

export const rootReducer = combineReducers({
  alerts,
  auth,
  // logIn,
  device,
  lng,
  structure,
  tech
})

export { startAlert, removeAlert } from './alerts'
export {
  setState as authSetState,
  signUpStart,
  signUpSuccess,
  signUpFail,
  logInStart,
  logInSuccess,
  logInFail,
  logOut,
  signUpModalClosed,
  logInModalClosed,
  authSelector
} from './auth'
export { deviceSelector, setDeviceSize, openModal, closeModal, setEditable } from './device'
export { lngSwitch, lngSelector } from './lng'
export {
  setState as structureSetState,
  structureSelector,
  structureStart,
  structureSuccess,
  structureFail
} from './structure'
export {
  techSelector,
  startInitLoading,
  initLoadingSuccess,
  startTechIn,
  techInSuccess,
  techInFail,
  startLngs,
  lngsSuccess,
  lngsFail,
  startI18n,
  i18nInitiated,
  i18nSuccess,
  i18nFail
} from './tech'
export default rootReducer
