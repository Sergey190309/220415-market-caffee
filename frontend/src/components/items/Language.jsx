import React, { useState, useEffect } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import PropTypes from 'prop-types'
import i18next from 'i18next'

import { Dropdown } from 'semantic-ui-react'
import { axiosCommonLng } from '../../api/apiClient'

import { techSelector, lngSelector, lngSwitch } from '../../redux/slices'
// import { setLngAction } from '../../redux/actions/lng';

export const onChange = (value, setActiveLng, dispatch, _i18next = i18next) => {
  // console.log('Lanaguage component, onChange, value ->', value);
  _i18next.changeLanguage(value) // Set language in i18next.
  axiosCommonLng(value) // Set language for API calls in request header.
  setActiveLng(value) // Set this component's state.
  dispatch(lngSwitch(value)) // Change language in application state state.
}

export const Language = ({ onChange, i18next }) => {
  const [activeLng, setActiveLng] = useState(i18next.language) // Active language
  const [availableLngs, setAvailableLngs] = useState([]) // availableLngs languages
  const dispatch = useDispatch()
  const { loaded } = useSelector(techSelector)
  const { lng } = useSelector(lngSelector)

  // console.log('component, Language, i18next.language ->', i18next.language)

  useEffect(() => {
    if (loaded) {
      // console.log('component, Languages, eseEffect(loaded), i18next.languages ->', i18next.languages)
      setAvailableLngs(
        i18next.languages.map(lng => ({
          key: lng,
          value: lng,
          // text: lng,
          flag: lng === 'en' ? 'uk' : lng
        }))
      )
      setActiveLng(i18next.language)
    }
  }, [loaded, i18next])

  useEffect(() => {
    if (lng !== activeLng) {
      setActiveLng(lng) // caried out once when initiated with lng 'ru'
    }
  }, [activeLng, lng])

  const _onChange = (evt, { value }) => {
    evt.preventDefault()
    onChange(value, setActiveLng, dispatch)
    // console.log('component, Language, _onChange, activeLng ->', activeLng)
  }
  // console.log('component, Language, activeLng ->', activeLng)

  return (
    <Dropdown
      name='langSwitcher'
      floating
      button
      defaultOpen={false}
      placeholder='langSwitcher'
      data-testid='dropdown'
      options={availableLngs}
      onChange={_onChange}
      value={activeLng}
    />
  )
}

Language.defaultProps = {
  onChange: onChange,
  i18next: i18next
}

Language.propTypes = {
  onChange: PropTypes.func.isRequired,
  i18next: PropTypes.object.isRequired
}

export default Language
