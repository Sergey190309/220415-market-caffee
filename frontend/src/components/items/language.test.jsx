import React from 'react'
import { Provider } from 'react-redux'
import { render, screen, waitFor } from '@testing-library/react'

import store from '../../redux/store'
import Language, { onChange } from './Language'
import { initLoadingSuccess } from '../../redux/slices'
import userEvent from '@testing-library/user-event'

describe('Language switcher testing', () => {
  describe('function testing', () => {
    test('onChange testing', () => {
      const mockValue = 'mockValue'
      const mockSetActiveLng = jest.fn()
      const mockDispatch = jest.fn()
      const mockI18next = {
        changeLanguage: jest.fn()
      }

      onChange(mockValue, mockSetActiveLng, mockDispatch, mockI18next)
      expect(mockSetActiveLng).toHaveBeenCalledTimes(1)
      expect(mockSetActiveLng).toHaveBeenCalledWith(mockValue)

      expect(mockDispatch).toHaveBeenCalledTimes(1)

      expect(mockI18next.changeLanguage).toHaveBeenCalledTimes(1)
      expect(mockI18next.changeLanguage).toHaveBeenCalledWith(mockValue)
    })
  })
  describe('Component', () => {
    const mockProps = {
      onChange: jest.fn(),
      i18next: {
        language: 'ru',
        languages: ['en', 'ru']
      }
    }

    test('it rendered normal way (snapshot)', () => {
      const { container, rerender } = render(
        <Provider store={store}>
          <Language {...mockProps} />
        </Provider>
      )
      store.dispatch(initLoadingSuccess())
      rerender(
        <Provider store={store}>
          <Language {...mockProps} />
        </Provider>
      )
      expect(container).toMatchSnapshot()
      // screen.debug();
    })

    describe('user action', () => {
      test('FAILED click, I am unable to simulate click', async () => {
        /**
         * It looks semantic react limitation:
         * https://stackoverflow.com/questions/52813527/cannot-check-expectelm-not-tobevisible-for-semantic-ui-react-component
         * Not sure about
         */
        store.dispatch(initLoadingSuccess())
        render(
          // const {rerender} = render(
          <Provider store={store}>
            <Language {...mockProps} />
          </Provider>
        )
        /**
         * Structure
         */
        const dropdown = screen.getByTestId('dropdown')
        expect(dropdown).toBeVisible()
        expect(dropdown.children).toHaveLength(3)

        const display = dropdown.children[0]
        expect(display).toHaveClass('divider text', { exact: true })
        const icon = dropdown.children[1]
        expect(icon).toHaveClass('dropdown icon', { exact: true })

        const options = dropdown.children[2]
        // expect(options).toHaveClass('menu transition', { exact: true });
        expect(options).toBeVisible()
        expect(options.children).toHaveLength(2)
        const option0 = options.children[0]
        expect(option0).toHaveClass('active selected item', { exact: true })
        const option1 = options.children[1]
        expect(option1).toHaveClass('item', { exact: true })

        /**
         * Actions
         */
        userEvent.click(dropdown)

        const clickedOptions = screen.getAllByRole('option') //
        expect(clickedOptions).toHaveLength(2)
        expect(clickedOptions[1]).toHaveClass('item', { exact: true })
        await waitFor(() => {
          userEvent.click(clickedOptions[1])
          const clickedOption = clickedOptions[1]
          expect(clickedOption).toHaveClass('item', { exact: true })
        })

        // rerender(
        //   <Provider store={store}>
        //     <Language {...mockProps} />
        //   </Provider>
        // )
        // screen.debug();
      })
    })
  })
})
