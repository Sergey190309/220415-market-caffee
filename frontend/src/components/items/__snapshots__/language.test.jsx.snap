// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Language switcher testing Component it rendered normal way (snapshot) 1`] = `
<div>
  <div
    aria-disabled="false"
    aria-expanded="false"
    class="ui button floating dropdown"
    data-testid="dropdown"
    name="langSwitcher"
    role="listbox"
    tabindex="0"
  >
    <div
      aria-atomic="true"
      aria-live="polite"
      class="divider text"
      role="alert"
    >
      <i
        class="uk flag"
      />
    </div>
    <i
      aria-hidden="true"
      class="dropdown icon"
    />
    <div
      class="menu transition"
    >
      <div
        aria-checked="true"
        aria-selected="true"
        class="active selected item"
        role="option"
        style="pointer-events: all;"
      >
        <i
          class="uk flag"
        />
      </div>
      <div
        aria-checked="false"
        aria-selected="false"
        class="item"
        role="option"
        style="pointer-events: all;"
      >
        <i
          class="ru flag"
        />
      </div>
    </div>
  </div>
</div>
`;
